### PR TITLE
Enable TypeScript strict mode

### LINKED ISSUE #350: Enable TypeScript strict mode
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/pageres/issues/350)

It would require removing this:

https://github.com/sindresorhus/pageres/blob/e54a5b3a68fb572e6347a11bad80800dcff2adc0/tsconfig.json#L10-L11

Then fixing all type errors and import the appropriate `@types/*` packages.

> There is a $30.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/pageres/issues/350).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/pageres/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by satyarohith**
We need definitions for the below modules before we proceed.
- [x] `easydate`
- [x] `array-differ`
- [x] `get-res`
- [x] `capture-website` see https://github.com/sindresorhus/capture-website/pull/19
- [x] `viewport-list`

closes #350

### CODE REVIEW

**Comment by satyarohith** [comment]
@sindresorhus Shall I just `declare module` for [`png-js`](https://github.com/devongovett/png.js#readme)? It is quite old and the documentation isn't adequate to proceed. We're using it for tests.

**Comment by sindresorhus** [comment]
`@types/png.js`?

**Comment by satyarohith** [comment]
No, I was talking about [`png-js`](https://npm.im/png-js).

**Comment by sindresorhus** [comment]
Shall I just declare module for png-js?

Yes

File: test/cookie.ts [hunk 1]
```diff
@@ -1,10 +1,15 @@
-import test from 'ava';
+import test, {ExecutionContext} from 'ava';
+// @ts-ignore
 import PNG from 'png-js';
```
Comment (satyarohith) [inline test/cookie.ts]: I couldn't declare the module with  `.d.ts` file because `ts-node` is throwing the below error.
```
TypeError: Unable to require `.d.ts` file.
  This is usually the result of a faulty configuration or import. Make sure there is a `.js`, `.json` or another executable extension and loader (attached before `ts-node`) available alongside `png-js.d.ts`
```
I would love to know if this can be solved in a better way.

File: test/cookie.ts [hunk 2]
```diff
@@ -1,10 +1,15 @@
...
 import pify from 'pify';
 import Pageres from '../source';
 import {createCookieServer} from './_server';
 
-async function cookieTest(input, t): Promise<void> {
+interface Cookie {
+	[key: string]: string;
+}
+
+async function cookieTest(input: string | Cookie, t: ExecutionContext): Promise<void> {
```
Comment (satyarohith) [inline test/cookie.ts]: The appropriate input type would be  `string | SetCookie`. I hacked it this way to avoid importing `SetCookie` from `puppeteer`.

File: package.json [hunk 3]
```diff
@@ -70,8 +70,23 @@
 	},
 	"devDependencies": {
 		"@sindresorhus/tsconfig": "^0.2.0",
+		"@types/array-uniq": "^1.0.1",
+		"@types/easydate": "^2.2.0",
```
Comment (sindresorhus) [inline package.json]: Note: https://github.com/sindresorhus/pageres/commit/8c3e0772e3ceed6e0423111f41f614951a39c534

File: package.json [hunk 4]
```diff
@@ -70,8 +70,23 @@
...
+		"@types/file-type": "^10.6.0",
```
Comment (sindresorhus) [inline package.json]: `file-type` has types built-in now.

File: package.json [hunk 5]
```diff
@@ -70,8 +70,23 @@
...
+		"@types/filenamify-url": "^1.0.0",
```
Comment (sindresorhus) [inline package.json]: This one also has it built-in now.

File: package.json [hunk 6]
```diff
@@ -70,8 +70,23 @@
...
+		"@types/get-res": "^3.0.1",
 		"@types/image-size": "^0.7.0",
+		"@types/lodash.template": "^4.4.5",
+		"@types/log-symbols": "^2.0.0",
+		"@types/make-dir": "^1.0.3",
```
Comment (sindresorhus) [inline package.json]: This one also has it built-in now.

File: package.json [hunk 7]
```diff
@@ -70,8 +70,23 @@
...
+		"@types/mem": "^1.1.2",
```
Comment (sindresorhus) [inline package.json]: This one also has it built-in now.

File: package.json [hunk 8]
```diff
@@ -70,8 +70,23 @@
...
 		"@types/node": "^10.12.23",
+		"@types/path-exists": "^3.0.0",
+		"@types/pify": "^3.0.2",
+		"@types/plur": "^3.0.0",
+		"@types/sinon": "^7.0.6",
+		"@types/unused-filename": "^1.0.0",
```
Comment (sindresorhus) [inline package.json]: This one also has it built-in now.

File: test/cookie.ts (hunk 1 above)
Comment (sindresorhus) [reply test/cookie.ts]: https://github.com/TypeStrong/ts-node/issues/615

File: package.json [hunk 9]
```diff
@@ -7,7 +7,7 @@
 	"author": {
 		"name": "Sindre Sorhus",
 		"email": "sindresorhus@gmail.com",
-		"url": "sindresorhus.com"
+		"url": "https://sindresorhus.com"
```
Comment (sindresorhus) [inline package.json]: Don't do unrelated changes.

Comment (satyarohith) [reply package.json]: This was to satisfy my editor haha. I reverted it. :)

File: test/cookie.ts (hunk 1 above)
Comment (satyarohith) [reply test/cookie.ts]: I have tried all mentioned solutions but none worked.

Comment (satyarohith) [reply test/cookie.ts]: I will try to submit `png-js` to DefinitelyTyped.

**Comment by satyarohith** [comment]
@sindresorhus Maybe a rerun might fix the tests.

**Comment by satyarohith** [comment]
Ping!

File: package.json [hunk 10]
```diff
@@ -70,9 +70,19 @@
...
 		"@types/image-size": "^0.7.0",
+		"@types/lodash.template": "^4.4.5",
+		"@types/log-symbols": "^2.0.0",
 		"@types/node": "^11.11.4",
+		"@types/path-exists": "^3.0.0",
+		"@types/pify": "^3.0.2",
+		"@types/plur": "^3.0.0",
 		"@types/png.js": "^0.2.0",
+		"@types/sinon": "^7.0.6",
+		"@types/viewport-list": "^5.1.1",
```
Comment (sindresorhus) [inline package.json]: Go through the packages here and check if some have gotten the TS types bundled now.

File: source/index.ts [hunk 11]
```diff
@@ -184,9 +184,9 @@ export default class Pageres extends EventEmitter {
 	successMessage(): void {
 		const {screenshots, sizes, urls} = this.stats;
 		const words = {
-			screenshots: plur('screenshot', screenshots),
-			sizes: plur('size', sizes),
-			urls: plur('url', urls)
+			screenshots: plur('screenshot', Number(screenshots)),
```
Comment (sindresorhus) [inline source/index.ts]: You're fixing this in the wrong place. `this.stats` should have the correct types.

File: test/_server.ts [hunk 12]
```diff
@@ -24,6 +24,7 @@ const baseCreateServer = (fn): (() => Promise<TestServer>) => {
 		server.url = `http://${host}:${port}`;
 		server.protocol = 'http';
 		server.listen(port);
+		// @ts-ignore
```
Comment (sindresorhus) [inline test/_server.ts]: Don't use `@ts-ignore`. Resolve the type issue instead. Applies to everywhere where this is used.

You could see https://github.com/sindresorhus/pageres/pull/371/files for inspiration.

File: tsconfig.json [hunk 13]
```diff
@@ -5,12 +5,9 @@
...
-		],
-
-		// TODO: Remove this
-		"strict": false
+		]
 	},
 	"include": [
 		"source"
-	]
+	],
```
Comment (sindresorhus) [inline tsconfig.json]: Don't do unrelated changes.

**Comment by sindresorhus** [comment]
Also upgrade to latest TypeScript and other dependencies to ensure the types are still valid.

File: tsconfig.json (hunk 13 above)
Comment (satyarohith) [reply tsconfig.json]: Sorry, that's my editor :sweat_smile:

**Comment by satyarohith** [comment]
@sindresorhus I have updated the PR. :)

**Comment by sindresorhus** [comment]
Looks good. Thanks :)
